components/tidb_query/src/expr/builtin_string.rs

```diff
@@ -225,6 +225,27 @@ impl ScalarFunc {
         }
     }
 
+    // see https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_repeat
+    #[inline]
+    pub fn repeat<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let val = try_opt!(self.children[0].eval_string(ctx, row));
+        let count = try_opt!(self.children[1].eval_int(ctx, row));
+        if count < 1 {
+            Ok(Some(Cow::Borrowed(b"")))
+        } else if count == 1 {
+            // return the input string when count is 1 to save one copy
+            Ok(Some(val))
+        } else {
+            // here count > 1, so convert it into usize should be ok
+            // consider using repeat() when rust/48784 is stablized
+            Ok(Some(Cow::Owned(vec![val; count as usize].concat())))
```

Comment:
In TiDB num is restricted to MaxInt32, maybe we can follow it:

```golang
	if num > math.MaxInt32 {
		num = math.MaxInt32
	}
```

Reply:
I'm also worried about the performance of creating a vector first then join elements. I don't know whether or not compiler is smart enough to optimize it. Prefer to do it in the naive way, i.e. manually loop and extend elements.

Reply:
1. Added the check.
2. Since TiKV has upgraded to a newer rust version, there is a `slice::repeat` in the std library. I think using that cloud be a better idea.